File: Sliding_Window/LC_239.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.28

class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        // C++ has the same data structure for deque!
        deque<int> q;
        vector<int> ans;
        for (int i = 0; i < nums.size(); i++){
            while (!q.empty() && q.back() < nums[i]){
                q.pop_back();
            }
            q.push_back(nums[i]);
            if (i-k+1 >= 0){
                ans.push_back(q.front());
                if (nums[i-k+1] == q.front()){
                    q.pop_front();
                }
            }
        }
        return ans;
    }
};

File: Sliding_Window/LC_239.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.28

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        q = collections.deque()
        ans = []
        for i in range(len(nums)):
            while q and q[-1] < nums[i]:
                q.pop()
            q.append(nums[i])
            if i >= k -1:
                ans.append(q[0])
                if q[0] == nums[i-k+1]:
                    q.popleft()
        return ans